sky/backends/provision_utils.py

```diff
@@ -0,0 +1,502 @@
+    display_name: str
+    name_on_cloud: str
+
+    def __repr__(self) -> str:
+        return repr(self.display_name)
+
+    def __str__(self) -> str:
+        return self.display_name
+
+
+@contextlib.contextmanager
+def _add_logger_handlers(log_path: str):
+    """Add file handler for logger."""
+    try:
+        log_abs_path = pathlib.Path(log_path).expanduser().absolute()
+        fh = logging.FileHandler(log_abs_path)
+        fh.setFormatter(sky_logging.FORMATTER)
+        fh.setLevel(logging.DEBUG)
+        logger.addHandler(fh)
+
+        # Redirect underlying provision logs to file.
+        provision_logger = logging.getLogger('sky.provision')
+        stream_handler = sky_logging.RichSafeStreamHandler(sys.stdout)
+        stream_handler.flush = sys.stdout.flush  # type: ignore
+        stream_handler.setFormatter(sky_logging.DIM_FORMATTER)
+        stream_handler.setLevel(logging.WARNING)
+
+        provision_logger.addHandler(fh)
+        provision_logger.addHandler(stream_handler)
+
+        provision_config.config.provision_log = log_abs_path
+        yield
+    finally:
+        logger.removeHandler(fh)
+        provision_logger.removeHandler(fh)
+        provision_logger.removeHandler(stream_handler)
+        stream_handler.close()
+        fh.close()
+
+
+def _bulk_provision(
+    cloud: clouds.Cloud,
+    region: clouds.Region,
+    zones: Optional[List[clouds.Zone]],
+    cluster_name: ClusterName,
+    bootstrap_config: provision_common.InstanceConfig,
+) -> provision_common.ProvisionMetadata:
+    provider_name = repr(cloud)
+    region_name = region.name
+
+    style = colorama.Style
+
+    if not zones:
+        # For Azure, zones is always an empty list.
+        zone_str = 'all zones'
+    else:
+        zone_str = ','.join(z.name for z in zones)
+
+    if isinstance(cloud, clouds.Local):
+        logger.info(f'{style.BRIGHT}Launching on local cluster '
+                    f'{cluster_name!r}.')
+    else:
+        logger.info(f'{style.BRIGHT}Launching on {cloud} '
+                    f'{region_name}{style.RESET_ALL} ({zone_str})')
+
+    start = time.time()
+    with rich_utils.safe_status('[bold cyan]Launching[/]') as status:
+        try:
+            # TODO(suquark): Should we cache the bootstrapped result?
+            #  Currently it is not necessary as bootstrapping takes
+            #  only ~3s, caching it seems over-engineering and could
+            #  cause other issues like the cache is not synced
+            #  with the cloud configuration.
+            config = provision.bootstrap_instances(provider_name, region_name,
+                                                   cluster_name.name_on_cloud,
+                                                   bootstrap_config)
+        except Exception as e:
+            logger.error('Failed to bootstrap configurations for '
+                         f'{cluster_name!r} with the following error:\n'
+                         f'{common_utils.format_exception(e)}')
```

Comment:
It might be a good idea to dim this part, just like the old behavior? Right now, a chunk of text is displayed together, which is hard to distinguish.

Reply:
Dimed the message. It will not fail the failover entirely.